flask/app.py

```diff
@@ -1029,6 +1029,27 @@ def index():
         rule = self.url_rule_class(rule, methods=methods, **options)
         rule.provide_automatic_options = provide_automatic_options
 
+        for prior_rule in self.url_map.iter_rules():
```

Comment:
Part of why I added this change to `add_url_rule` is because this function holds the assert that makes sure no function name is used twice.

If you still think the modification should be moved to `WerkZeug.Map` should I follow the same tuple check?